File: src/lib/GemInterface.py
import os
import asyncio
from dotenv import load_dotenv
import requests
from requests.adapters import HTTPAdapter
from urllib3.util.retry import Retry
from bs4 import BeautifulSoup
from typing import Any, Optional, AsyncIterator
import json

from ollama import chat, AsyncClient
from ollama import ChatResponse

class AiInterface:
    """
    AI Interface using Ollama for local LLM inference with streaming support.
    Keeps the original synchronous web scraper (requests + BeautifulSoup) but improves it
    by adding a requests.Session with browser-like headers and a retry strategy to reduce
    403/429/5xx failures. Everything else remains async-friendly by running blocking
    operations in a threadpool.

    Usage:
      ai = AiInterface()
      result = asyncio.run(ai.Archie("When is fall break?"))
    """

    def __init__(
        self,
        debug: bool = False,
        scraper_max_retries: int = 3,
        scraper_backoff_factor: float = 1.0,
        scraper_timeout: int = 15,
    ):
        # Load the variables from the .env file into the environment
        load_dotenv()

        # Retrieve the model name from environment (defaults to llama2 if not set)
        self.model = os.getenv("MODEL", "llama2")
        
        # Initialize Ollama async client for streaming
        self.async_client = AsyncClient()

        # Debug flag
        self.debug = debug

        # Scraper configuration
        self.scraper_timeout = scraper_timeout

        # Create a requests.Session with headers and retry strategy
        self.session: requests.Session = requests.Session()
        # Browser-like default headers to reduce likelihood of 403 from simple bot checks
        self.session.headers.update({
            "User-Agent": (
                "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
                "AppleWebKit/537.36 (KHTML, like Gecko) "
                "Chrome/120.0.0.0 Safari/537.36"
            ),
            "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
            "Accept-Language": "en-US,en;q=0.9",
        })

        # Configure retries for transient errors and common rate-limit statuses.
        retry_strategy = Retry(
            total=scraper_max_retries,
            status_forcelist=[429, 500, 502, 503, 504],
            allowed_methods=["HEAD", "GET", "OPTIONS"],
            backoff_factor=scraper_backoff_factor,
            raise_on_status=False,
        )
        adapter = HTTPAdapter(max_retries=retry_strategy)
        self.session.mount("https://", adapter)
        self.session.mount("http://", adapter)

    def _log(self, *args):
        if self.debug:
            print("[AiInterface DEBUG]", *args)

    def generate_text(self, prompt: str, system_prompt: str = "") -> str:
        """
        Synchronous method: generate text using Ollama client.
        This method is kept for backwards compatibility.
        """
        try:
            messages = []
            if system_prompt:
                messages.append({
                    'role': 'system',
                    'content': system_prompt,
                })
            messages.append({
                'role': 'user',
                'content': prompt,
            })
            
            response: ChatResponse = chat(model=self.model, messages=messages)
            return response['message']['content']
        except Exception as e:
            self._log(f"Error during text generation: {e}")
            return "An error occurred during text generation"

    def scrape_website(self, url: str, timeout: Optional[int] = None) -> str:
        """
        Improved synchronous web scraper that:
        - uses a persistent requests.Session with browser-like headers
        - has a Retry strategy for transient status codes (429, 5xx)
        - keeps the interface synchronous (requests + BeautifulSoup) as requested
        """
        try:
            to = timeout if timeout is not None else self.scraper_timeout
            self._log(f"Scraping {url} with timeout={to}")
            response = self.session.get(url, timeout=to, allow_redirects=True)
            # If raise_for_status raises, we catch below and return a helpful message
            try:
                response.raise_for_status()
            except requests.HTTPError as http_err:
                # Provide helpful debug string but still try to return body if available
                self._log(f"HTTP error for {url}: {http_err} (status {response.status_code})")
                # If server returned some HTML (e.g., Cloudflare block), parse and return its text
                # Otherwise, return the error string
                if response.text:
                    soup = BeautifulSoup(response.text, "html.parser")
                    return soup.get_text()
                return f"HTTP error when scraping {url}: {http_err}"

            soup = BeautifulSoup(response.text, 'html.parser')
            return soup.get_text()
        except requests.RequestException as e:
            self._log(f"RequestException when scraping {url}: {e}")
            return "An error occurred while scraping the website"
        except Exception as e:
            self._log(f"Unexpected error when scraping {url}: {e}")
            return "An unexpected error occurred while scraping the website"


    async def generate_text_async(self, prompt: str, system_prompt: str = "") -> str:
        """
        Async wrapper around Ollama chat for non-streaming responses.
        Runs the operation in a threadpool to avoid blocking the event loop.
        """
        def _generate_text(prompt: str, system_prompt: str = "") -> str:
            messages = []
            if system_prompt:
                messages.append({
                    'role': 'system',
                    'content': system_prompt,
                })
            messages.append({
                'role': 'user',
                'content': prompt,
            })
            
            try:
                response: ChatResponse = chat(model=self.model, messages=messages)
                return response['message']['content']
            except Exception as e:
                self._log(f"Error during text generation: {e}")
                return "An error occurred during text generation"
        
        loop = asyncio.get_running_loop()
        return await loop.run_in_executor(None, _generate_text, prompt, system_prompt)
    
    async def generate_text_streaming(self, prompt: str, system_prompt: str = "") -> AsyncIterator[str]:
        """
        Async streaming generator that yields tokens as they are generated by Ollama.
        This allows for real-time display of the AI's thinking process.
        
        Usage:
            async for token in ai.generate_text_streaming(prompt, system):
                print(token, end='', flush=True)
        """
        messages = []
        if system_prompt:
            messages.append({
                'role': 'system',
                'content': system_prompt,
            })
        messages.append({
            'role': 'user',
            'content': prompt,
        })
        
        try:
            stream = await self.async_client.chat(
                model=self.model,
                messages=messages,
                stream=True
            )
            
            async for chunk in stream:
                if 'message' in chunk and 'content' in chunk['message']:
                    yield chunk['message']['content']
        except Exception as e:
            with open("error.txt", "w", encoding="utf-8") as f:
                f.write(str(e))
            yield "An error occurred during streaming"

    async def _run_in_executor(self, func: Any, *args, **kwargs):
        """
        Helper to run any synchronous function in the default threadpool and return its result.
        """
        loop = asyncio.get_running_loop()
        return await loop.run_in_executor(None, lambda: func(*args, **kwargs))
    
    async def Archie(self, query: str) -> str:
        """
        Main async entry point for the Archie AI assistant.
        Uses scraped data from JSON file to provide context for answering queries.
        """
        with open("data/scrape_results.json", "r", encoding="utf-8") as f:
            results = json.load(f)
        
        system_prompt = f"""You are ArchieAI, an AI assistant for Arcadia University. You are here to help students, faculty, and staff with any questions they may have about the university.

You are made by students for a final project. You must be factual and accurate based on the information provided. It is ok to say "I don't know" if you are unsure.

Respond based on your knowledge up to 2025.

Use the following content to better answer the query:
{json.dumps(results, indent=2)}"""

        return await self.generate_text_async(query, system_prompt=system_prompt)
    
    async def Archie_streaming(self, query: str) -> AsyncIterator[str]:
        """
        Streaming version of Archie that yields tokens as they are generated.
        This provides a better user experience by showing the AI "thinking" in real-time.
        
        Usage:
            async for token in ai.Archie_streaming("When is fall break?"):
                print(token, end='', flush=True)
        """
        with open("data/scrape_results.json", "r", encoding="utf-8") as f:
            results = json.load(f)
        
        system_prompt = f"""You are ArchieAI, an AI assistant for Arcadia University. You are here to help students, faculty, and staff with any questions they may have about the university.

You are made by students for a final project. You must be factual and accurate based on the information provided. It is ok to say "I don't know" if you are unsure.

Respond based on your knowledge up to 2025.

Use the following content to better answer the query:
{json.dumps(results, indent=2)}"""

        async for token in self.generate_text_streaming(query, system_prompt=system_prompt):
            yield token
    

if __name__ == "__main__":
    # Simple test of the AiInterface
    ai = AiInterface(debug=True)
    test_query = "When is fall break?"
    response = asyncio.run(ai.Archie(test_query))
    print("Response to query:", test_query)
    print(response)

File: src/lib/GemInterface.rs
// src/lib/GemInterface.rs

use dotenv::dotenv;
use std::env;
use std::error::Error;
use std::fmt;

// Imports needed for the successful version
use gemini_rust::{
    Message, Part, Role,
    client::{Error as ClientError, Gemini},
};

// Required for synchronous calling of the async Gemini client
use futures::executor::block_on;

// Crates for web scraping (HTTP and HTML parsing)
use reqwest::{Client, Error as ReqwestError};
use scraper::{Html, Selector};

// ----------------------------------------------------------------------
// 1. Error Handling Definition (Cleaned up for no conflict)
// ----------------------------------------------------------------------

/// Custom error type to encompass potential issues.
#[derive(Debug)]
pub enum AiInterfaceError {
    Environment(String),
    // 🟢 FIX: Renaming to 'Api' to reflect that it handles all boxed errors from the API side.
    Api(Box<dyn Error + Send + Sync>),
    Request(String), // Changed to String as we can't reliably convert reqwest::Error specifically
    General(String),
}

impl fmt::Display for AiInterfaceError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            AiInterfaceError::Environment(msg) => write!(f, "Environment Error: {}", msg),
            AiInterfaceError::Api(e) => write!(f, "Gemini API Error: {}", e),
            AiInterfaceError::Request(msg) => write!(f, "Request Error: {}", msg),
            AiInterfaceError::General(msg) => write!(f, "General Error: {}", msg),
        }
    }
}

impl Error for AiInterfaceError {}

// 🟢 Generic conversion for all errors satisfying the bounds (includes ClientError and Gemini builder errors)
impl<T: Error + Send + Sync + 'static> From<T> for AiInterfaceError
where
    ClientError: From<T>, // Constraint to errors likely from the Gemini crate
{
    fn from(error: T) -> Self {
        // All boxed errors go into the API variant
        AiInterfaceError::Api(Box::new(error))
    }
}

// 🔴 NOTE: We must remove impl From<reqwest::Error> to resolve the conflict.
// We handle reqwest::Error manually in the function body instead of using the '?' operator.

// ----------------------------------------------------------------------
// 2. AiInterface Struct and Methods
// ----------------------------------------------------------------------

/// The main structure for interacting with the Gemini API and web scraping.
pub struct AiInterface {
    client: Gemini,
    http_client: reqwest::Client,
}

impl AiInterface {
    /// Constructor: Initializes the client and loads environment variables.
    pub fn new() -> Result<Self, AiInterfaceError> {
        dotenv().ok();

        let api_key = env::var("GEMINI_API_KEY").map_err(|_| {
            AiInterfaceError::Environment(
                "GEMINI_API_KEY not found. Check your .env file or environment.".to_string(),
            )
        })?;

        // Uses the '?' operator, handled by the generic From<T> impl.
        let client = Gemini::new(api_key)?;
        let http_client = reqwest::Client::new();
        Ok(AiInterface {
            client,
            http_client,
        })
    }

    /// Generate text based on the given prompt using a Gemini model.
    pub async fn generate_text(&self, prompt: &str) -> Result<String, AiInterfaceError> {
        let message = Message::user(prompt.to_string());
        let messages = vec![message];

        let builder = self
            .client
            .generate_content()
            .with_model_message("gemini-2.5-flash-lite")
            .with_messages(messages);

        // Execute the builder. The '?' operator uses the generic From<T> impl.
        let response = builder.execute().await?;
        let text = response
            .candidates
            .first()
            .and_then(|c| c.content.parts.as_ref())
            .and_then(|parts_vec| parts_vec.first())
            .and_then(|p| match p {
                Part::Text { text, .. } => Some(text.clone()),
                _ => None,
            })
            .unwrap_or_else(|| "The model returned no text.".to_string());

        Ok(text)
    }

    /// Scrape the content of a given website and return it as a string.
    pub async fn scrape_website(&self, url: &str) -> Result<String, AiInterfaceError> {
        // 🟢 FIX: Manual error handling for reqwest since the '?' operator is blocked.
        match self.http_client.get(url).send().await {
            Ok(response) => match response.error_for_status() {
                Ok(clean_response) => {
                    let html_content = match clean_response.text().await {
                        Ok(text) => text,
                        Err(e) => {
                            return Err(AiInterfaceError::Request(format!(
                                "Failed to read response body: {}",
                                e
                            )));
                        }
                    };

                    let document = Html::parse_document(&html_content);

                    let selector = match Selector::parse("body") {
                        Ok(s) => s,
                        Err(_) => {
                            return Err(AiInterfaceError::General(
                                "Failed to parse selector".to_string(),
                            ));
                        }
                    };

                    let text_content: String = document
                        .select(&selector)
                        .flat_map(|element| element.text())
                        .collect::<String>()
                        .trim()
                        .to_string();

                    Ok(text_content)
                }
                Err(e) => Err(AiInterfaceError::Request(format!(
                    "HTTP error status: {}",
                    e
                ))),
            },
            Err(e) => Err(AiInterfaceError::Request(format!(
                "Failed to connect or send request: {}",
                e
            ))),
        }
    }

    /// Generate a response for a query using Arcadia University website content.
    pub async fn archie(&self, query: &str) -> Result<String, AiInterfaceError> {
        let website = self.scrape_website("https://www.arcadia.edu/").await?;
        let events = self.scrape_website("https://www.arcadia.edu/events/?mode=month").await?;
        let about = self.scrape_website("https://www.arcadia.edu/about-arcadia/").await?;

        let prompt = format!(
            "System: You are ArchieAI an AI assistant for Arcadia University. You are here to help students, faculty, and staff with any questions they may have about the university. You were made by Eva Akselrad and Ab.
            Using the following website content, answer the query: {}\n\nWebsite Content:\n{}\n\nEvents Content:\n{}\n\nAbout Content:\n{}",
            query, website, events, about
        );

        self.generate_text(&prompt).await
    }
}
